tests/compile-fail/eval_order_dependence.rs

```diff
@@ -0,0 +1,53 @@
+#![feature(plugin)]
+#![plugin(clippy)]
+
+#[deny(eval_order_dependence)]
+#[allow(unused_assignments, unused_variables, many_single_char_names, no_effect, dead_code, blacklisted_name)]
+fn main() {
+    let mut x = 0;
+    let a = { x = 1; 1 } + x;
+                        //~^ ERROR unsequenced read
+
+    // Example from iss#277
+    x += { x = 20; 2 }; //~ERROR unsequenced read
+
+    // Does it work in weird places?
+    // ...in the base for a struct expression?
+    struct Foo { a: i32, b: i32};
+    let base = Foo { a: 4, b: 5 };
+    let foo = Foo { a: x, .. { x = 6; base } };
+                    //~^ ERROR unsequenced read
+    // ...inside a closure?
+    let closure = || {
+        let mut x = 0;
+        x += { x = 20; 2}; //~ERROR unsequenced read
+    };
+    // ...not across a closure?
+    let mut y = 0;
+    let b = (y, || { y = 1 });
+
+    // && and || evaluate left-to-right.
+    let a = { x = 1; true } && (x == 3);
+    let a = { x = 1; true } || (x == 3);
+
+    // Make sure we don't get confused by alpha conversion.
+    let a = { let mut x = 1; x = 2; 1} + x;
+
+    // No warning if we don't read the variable...
+    x = { x = 20; 2 };
```

Comment:
It's easy if you want to detect an assignments to `x` whose RHS also assigns to `x`; the lint would just need to put [here](https://github.com/Manishearth/rust-clippy/pull/1158/commits/4de996cb69855a63423ec958e21f088535d6f53c#diff-73cc6e5470cdf00577bce07ce2d8e946R204). I can add that if you want, but I don't know what to call it.